project_earned_value_task/models/project.py

```diff
@@ -0,0 +1,472 @@
+        except ZeroDivisionError:
+            res['etc'] = 0
+
+        # EAC: Estimate At Completion
+        try:
+            res['eac'] = res['bac'] / res['cpi']
+        except ZeroDivisionError:
+            res['eac'] = 0
+
+        # POC - Percent of Completion
+        try:
+            res['poc'] = res['ev'] / res['bac'] * 100
+        except ZeroDivisionError:
+            res['poc'] = 0
+
+        return res
+
+    def _get_earliest_latest_dates(self):
+        project = self
+        cr = self.env.cr
+        # Get the earliest and latest dates for the associated tasks
+        cr.execute("""SELECT MIN(date_start)
+                   FROM project_task
+                   WHERE project_id=%s""", (project.id, ))
+        min_date_start = cr.fetchone()[0] or 0
+        if min_date_start == 0:
+            datetime_start = datetime.today()
+        else:
+            datetime_start = datetime.strptime(min_date_start,
+                                               "%Y-%m-%d %H:%M:%S")
+
+        cr.execute("""SELECT MAX(date_end)
+                   FROM project_task
+                   WHERE project_id=%s""", (project.id, ))
+        max_date_end = cr.fetchone()[0] or 0
+        if max_date_end == 0:
+            datetime_end = datetime.today()
+        else:
+            datetime_end = datetime.strptime(max_date_end,
+                                             "%Y-%m-%d %H:%M:%S")
+
+        cr.execute("""SELECT MAX(date_last_stage_update)
+                   FROM project_task
+                   WHERE project_id=%s""", (project.id, ))
+        max_date_end_2 = cr.fetchone()[0] or 0
+        if max_date_end_2 == 0:
+            datetime_end_2 = datetime.today()
+        else:
+            datetime_end_2 = datetime.strptime(max_date_end,
+                                             "%Y-%m-%d %H:%M:%S")
+        datetime_end = max(datetime_end, datetime_end_2)
+
+        return datetime_start, datetime_end
+
+    def _get_budget_at_completion(self):
+        cr = self.env.cr
+        # Compute the Budget at Completion
+        cr.execute("""
+        SELECT SUM(PT.planned_hours * ip.value_float)
+        FROM project_task as PT
+        INNER JOIN resource_resource RES
+        ON RES.user_id = PT.user_id
+        INNER JOIN hr_employee EMP
+        ON EMP.resource_id = RES.id
+        INNER JOIN product_product PR
+        ON PR.id = EMP.product_id
+        INNER JOIN product_template PRT
+        ON PRT.id = PR.product_tmpl_id
+        LEFT JOIN ir_property ip
+        ON (ip.name='standard_price'
+        AND ip.res_id=CONCAT('product.template,',PRT.id)
+        AND ip.company_id=PT.company_id)
+        WHERE PT.project_id = %s""", (self.id, ))
+        res = cr.fetchone()
+        if res:
+            return res[0] or 0.0
+        else:
+            return 0.0
+
+    def _get_plan_cost_to_date(self):
+        cr = self.env.cr
+        today = datetime.today().strftime('%Y-%m-%d %H:%M:%S')
```

Comment:
What if we want to measure this at a different point in time, such as end of last month when preparing reporting?
I suggest accepting an additional optional argument with the date.